File: frontend/src/parser.ts
import type {
  Cue,
  LineError,
  ParseErrorCode,
  ParseResponse,
} from "./types";

// Client-side mirror of backend/app/parser.py: gives instant feedback and is
// covered by Vitest. The FastAPI /api/parse remains the authority — an import
// is only committed after the server accepts it.
const ERROR_MESSAGES: Record<ParseErrorCode, string> = {
  EMPTY_DOCUMENT: "导入内容为空",
  EMPTY_LINE: "空行：每行必须为“字幕文本|整数毫秒”",
  MISSING_SEPARATOR: "缺少分隔符 |：每行必须为“字幕文本|整数毫秒”",
  EMPTY_SUBTITLE: "字幕文本不得为空",
  INVALID_TIME: "时间必须为整数毫秒",
  NEGATIVE_TIME: "时间不得为负",
  NOT_STRICTLY_INCREASING: "各行时间必须严格递增、不得重复",
  REHEARSAL_IN_PROGRESS:
    "联排进行中：请先结束联排再重新导入计划，本次操作未改动已记录的敲击",
};

const INTEGER_RE = /^[+-]?\d+$/;

function lineError(line: number, code: ParseErrorCode): LineError {
  return { line, code, message: ERROR_MESSAGES[code] };
}

export function parseSchedule(raw: string): ParseResponse {
  const errors: LineError[] = [];
  const cues: Cue[] = [];

  if (raw.trim() === "") {
    return { valid: false, errors: [lineError(1, "EMPTY_DOCUMENT")], cues: [] };
  }

  let previousTime: bigint | null = null;

  raw.split(/\r\n|\r|\n/).forEach((rawLine, i) => {
    const lineNumber = i + 1;
    const line = rawLine.trim();

    if (line === "") {
      errors.push(lineError(lineNumber, "EMPTY_LINE"));
      return;
    }
    if (!line.includes("|")) {
      errors.push(lineError(lineNumber, "MISSING_SEPARATOR"));
      return;
    }

    const separator = line.lastIndexOf("|");
    const text = line.slice(0, separator).trim();
    const timeField = line.slice(separator + 1).trim();
    const lineErrors: LineError[] = [];

    if (text === "") {
      lineErrors.push(lineError(lineNumber, "EMPTY_SUBTITLE"));
    }

    let timeMs: bigint | null = null;
    if (!INTEGER_RE.test(timeField)) {
      lineErrors.push(lineError(lineNumber, "INVALID_TIME"));
    } else {
      const value = BigInt(timeField);
      if (value < 0n) {
        lineErrors.push(lineError(lineNumber, "NEGATIVE_TIME"));
      } else {
        timeMs = value;
      }
    }

    if (timeMs !== null) {
      if (previousTime !== null && timeMs <= previousTime) {
        lineErrors.push(lineError(lineNumber, "NOT_STRICTLY_INCREASING"));
      }
      previousTime = timeMs;
    }

    errors.push(...lineErrors);
    if (lineErrors.length === 0 && timeMs !== null) {
      cues.push({ text, time_ms: timeMs });
    }
  });

  if (errors.length > 0 || cues.length === 0) {
    return { valid: false, errors, cues: [] };
  }
  return { valid: true, errors: [], cues };
}

File: backend/app/parser.py
"""Schedule text parser.

Input format: one cue per line, ``subtitle text|integer milliseconds``.

Rules enforced here:
* subtitle text must not be empty;
* the time must be an integer and must not be negative;
* cue times must be strictly increasing (duplicates rejected).

Every error is located at its 1-based line number; several errors may be
reported for the same line. Whitespace-only documents are rejected.
"""
from __future__ import annotations

import re

from pydantic import BaseModel, Field

from .models import ERROR_MESSAGES, ParseErrorCode

_INTEGER_RE = re.compile(r"^[+-]?\d+$")


class Cue(BaseModel):
    text: str
    time_ms: int = Field(ge=0)


class LineError(BaseModel):
    line: int = Field(ge=1)
    code: ParseErrorCode
    message: str


class ParseResponse(BaseModel):
    valid: bool
    errors: list[LineError] = Field(default_factory=list)
    cues: list[Cue] = Field(default_factory=list)


def _line_error(line: int, code: ParseErrorCode) -> LineError:
    return LineError(line=line, code=code, message=ERROR_MESSAGES[code])


def parse_schedule(raw: str) -> ParseResponse:
    """Parse the raw schedule text and return either cues or located errors."""
    errors: list[LineError] = []
    parsed: list[tuple[int, Cue]] = []

    if raw is None or raw.strip() == "":
        return ParseResponse(
            valid=False,
            errors=[_line_error(1, ParseErrorCode.EMPTY_DOCUMENT)],
        )

    lines = raw.splitlines()
    previous_time: int | None = None
    previous_line: int | None = None

    for index, raw_line in enumerate(lines, start=1):
        line = raw_line.strip()
        if line == "":
            errors.append(_line_error(index, ParseErrorCode.EMPTY_LINE))
            continue

        if "|" not in line:
            errors.append(_line_error(index, ParseErrorCode.MISSING_SEPARATOR))
            continue

        text, _, time_field = line.rpartition("|")
        text = text.strip()
        time_field = time_field.strip()

        line_errors: list[LineError] = []
        if text == "":
            line_errors.append(_line_error(index, ParseErrorCode.EMPTY_SUBTITLE))

        time_ms: int | None = None
        if not _INTEGER_RE.match(time_field):
            line_errors.append(_line_error(index, ParseErrorCode.INVALID_TIME))
        else:
            value = int(time_field)
            if value < 0:
                line_errors.append(_line_error(index, ParseErrorCode.NEGATIVE_TIME))
            else:
                time_ms = value

        if time_ms is not None:
            if previous_time is not None and time_ms <= previous_time:
                order_error = _line_error(
                    index, ParseErrorCode.NOT_STRICTLY_INCREASING
                )
                assert previous_line is not None
                line_errors.append(order_error)
            previous_time = time_ms
            previous_line = index

        errors.extend(line_errors)
        if not line_errors and time_ms is not None:
            parsed.append((index, Cue(text=text, time_ms=time_ms)))

    if errors:
        return ParseResponse(valid=False, errors=errors)

    cues = [cue for _, cue in parsed]
    if not cues:
        return ParseResponse(
            valid=False,
            errors=[_line_error(1, ParseErrorCode.EMPTY_DOCUMENT)],
        )
    return ParseResponse(valid=True, cues=cues)
